Library/Homebrew/cmd/upgrade.rb

```diff
@@ -171,4 +200,60 @@ def upgrade_formula(f)
       nil
     end
   end
+
+  def reinstall_formula_from_source(f)
```

Comment:
Eek, there's rather a bit more here than I expected. I wonder how best to slim this down.

Reply:
Thought you'd say that 🤣 
I'd say including `cmd/reinstall` and add a parameter to `reinstall_formula` would be a better choice. Factoring out the reinstall logic into a new module like `LinkageChecker`-`cmd/linkage` and calling it from there would also work.

Reply:
> I'd say including cmd/reinstall and add a parameter to reinstall_formula would be a better choice. Factoring out the reinstall logic into a new module like LinkageChecker-cmd/linkage and calling it from there would also work.

@amyspark Yeh, I think so; having a new `Reinstall` class which both `cmd/reinstall` and `cmd/upgrade` can call would be ideal given `build_from_source` seems to be the only required new argument there.

Reply:
I've just sent in the changes, please check if it's OK for you.